core/engine/src/object/property_map.rs

```diff
@@ -108,42 +126,170 @@ impl IndexedProperties {
 
     /// Inserts a property descriptor with the specified key.
     fn insert(&mut self, key: u32, property: PropertyDescriptor) -> bool {
-        let vec = match self {
-            Self::Sparse(map) => return map.insert(key, property).is_some(),
-            Self::Dense(vec) => {
-                let len = vec.len() as u32;
-                if key <= len
-                    && property.value().is_some()
-                    && property.writable().unwrap_or(false)
-                    && property.enumerable().unwrap_or(false)
-                    && property.configurable().unwrap_or(false)
-                {
-                    // Fast Path: continues array access.
-
-                    let mut value = property
-                        .value()
-                        .cloned()
-                        .expect("already checked that the property descriptor has a value field");
-
-                    // If the key is pointing one past the last element, we push it!
-                    //
-                    // Since the previous key is the current key - 1. Meaning that the elements are continuos.
-                    if key == len {
-                        vec.push(value);
-                        return false;
+        let mut map =
+            match self {
+                Self::DenseI32(vec) => {
+                    let len = vec.len() as u32;
+                    if key <= len
+                        && property.value().is_some()
+                        && property.writable().unwrap_or(false)
+                        && property.enumerable().unwrap_or(false)
+                        && property.configurable().unwrap_or(false)
```

Comment:
Can we extract this check outside of the match? Maybe with a let else guard, so that we don't need to `expect` below on all branches.